Exercise 1 solution

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 27.8 MB/s 
     |████████████████████████████████| 49 kB 4.7 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
# create a model
model1 = ConcreteModel()

In [4]:
# declare decision variables
model1.x1 = Var(domain=Reals)
model1.x2 = Var(domain=Reals)
model1.x3 = Var(domain=Reals)
model1.x4 = Var(domain=Reals)
model1.x5 = Var(domain=Reals)
model1.x6 = Var(domain=Reals)
model1.x7 = Var(domain=Reals)

In [5]:
# declare objective
model1.cost = Objective(expr = -3*model1.x1 - 2*model1.x2 + 8*model1.x3 - 7*model1.x4 + 0.19*model1.x5 +11*model1.x6 + 20.2*model1.x7, sense=minimize)

In [6]:
# declare constraints
model1.constr1 = Constraint(expr = model1.x1 + model1.x2 + model1.x3 + model1.x4 + model1.x5 + model1.x6 + model1.x7 <= 256)
model1.constr2 = Constraint(expr = model1.x1 + model1.x2 + model1.x3 + model1.x4 + model1.x5 + model1.x6 + model1.x7 >= 256)

model1.constr3 = Constraint(expr = -1*model1.x2 + model1.x3 >= 10)
model1.constr4 = Constraint(expr = -3*model1.x1 + model1.x5 - 12*model1.x6 <= 105)
model1.constr5 = Constraint(expr = 18.5*model1.x2 + 30.5*model1.x4 - model1.x7 >= 28)
model1.constr6 = Constraint(expr = -1*model1.x3 - 4*model1.x5 + model1.x7 >= 31)
model1.constr7 = Constraint(expr = model1.x1 >= -5)
model1.constr8 = Constraint(expr = model1.x2 >= 0)
model1.constr9 = Constraint(expr = model1.x3 >= 5)
model1.constr10 = Constraint(expr = model1.x4 >= -3)
model1.constr11 = Constraint(expr = model1.x5 >= -4)
model1.constr12 = Constraint(expr = model1.x6 >= 0)
model1.constr13 = Constraint(expr = model1.x7 >= 10)

In [7]:
model1.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [8]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155673 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [9]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1010.76
  Upper bound: -1010.76
  Number of objectives: 1
  Number of constraints: 14
  Number of variables: 8
  Number of nonzeros: 33
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011990070343017578
# ----------------------------------------------------------
#   Solution Information
# ----------------------------

In [10]:
# display solution
print('/nExercise1 (3)')
print('\nCost = ', model1.cost())

print('\nDecision Variables')
print('x1 = ', model1.x1())
print('x2 = ', model1.x2())
print('x3 = ', model1.x3())
print('x4 = ', model1.x4())
print('x5 = ', model1.x5())
print('x6 = ', model1.x6())
print('x7 = ', model1.x7())

print('\nConstraints')
print('Constraint 1  = ', model1.constr1())
print('Constraint 2  = ', model1.constr2())
print('Constraint 3  = ', model1.constr3())
print('Constraint 4  = ', model1.constr4())
print('Constraint 5  = ', model1.constr5())
print('Constraint 6  = ', model1.constr6())
print('Constraint 7  = ', model1.constr7())
print('Constraint 8  = ', model1.constr8())
print('Constraint 9  = ', model1.constr9())
print('Constraint 10  = ', model1.constr10())
print('Constraint 11 = ', model1.constr11())
print('Constraint 12 = ', model1.constr12())
print('Constraint 13  = ', model1.constr13())

/nExercise1 (3)

Cost =  -1010.760000000002

Decision Variables
x1 =  -5.0
x2 =  0.0
x3 =  10.0
x4 =  230.0
x5 =  -4.0
x6 =  0.0
x7 =  24.9999999999999

Constraints
Constraint 1  =  255.9999999999999
Constraint 2  =  255.9999999999999
Constraint 3  =  10.0
Constraint 4  =  11.0
Constraint 5  =  6990.0
Constraint 6  =  30.9999999999999
Constraint 7  =  -5.0
Constraint 8  =  0.0
Constraint 9  =  10.0
Constraint 10  =  230.0
Constraint 11 =  -4.0
Constraint 12 =  0.0
Constraint 13  =  24.9999999999999


Exercise1 (4)-> 

constraint1 

x1 + x2 + x3 + x4 + x5 + x6 + x7 = 256 is Active 
because L.H.S is -5+0+10+230-4+0+24 = 256 = R.H.S

constraint3

-x2+x3 >=10 is Active
because L.H.S is -(0)+(10) = 10 = R.H.S

constraints4

-3x1+x5-12x6<=105  is Inactive
because L.H.S is -3(-5)+(-4)-12(0) = 11 != R.H.S

constraints5

18.5x2+30.5x4-x7 >= 28 is inactive
because L.H.S is 18.5(0)+30.5(230)-(-5) = 6990 != R.H.S

constraints6

-x3-4x5+x7>= 31 is Active
because L.H.S is -(10)-4(-4)+(25) = 31 = R.H.S


Exercise 2 Solution

$ \max \ \ 0.08x_1+0.036x_2+0.0156x_3+0.033x_4+0.054x_5+0.0125y \\ \textbf{s.t.} \ \ 3x_1-3x_2+5x_3-x_4+x_5-y\leq 180 \\ \ \ \ x_3+2x_4+4y\geq 30 \\ \ \ 3x_1+3.6x_2-1.5x_3+5.6x_4-1.8x_5+6x_6<=0 \\ \ 7x_1-2x_2-3x_3+4x_4+5x_5-6x_6+12y<=0 \\ \ x_1,x_2,x_3,x_4,x_5,x_6,\geq 0 \\ \ y\in \ [-0.2,5.5] $

In [11]:
!pip install -q pyomo

In [12]:
from pyomo.environ import *

In [13]:
model2 = ConcreteModel()

In [14]:
model2.x1 = Var(within=NonNegativeReals)
model2.x2 = Var(within=NonNegativeReals)
model2.x3 = Var(within=NonNegativeReals)
model2.x4 = Var(within=NonNegativeReals)
model2.x5 = Var(within=NonNegativeReals)
model2.x6 = Var(within=NonNegativeReals)
model2.y = Var(domain=Reals)

In [15]:
model2.cost = Objective(expr = 0.082*model2.x1 + 0.036*model2.x2 + 0.0156*model2.x3 + 0.033*model2.x4 + 0.0125*model2.y, sense=maximize)

In [16]:
model2.constr1 = Constraint(expr = 3*model2.x1 - 3*model2.x2 + 5*model2.x3 - model2.x4 + model2.x5 - model2.y <= 80)
model2.constr2 = Constraint(expr = model2.x3 + 2*model2.x4 + 4*model2.y >= 30)

model2.constr3 = Constraint(expr = 3*model2.x1 + 3.6*model2.x2 -1.5*model2.x3 + 5.6*model2.x4 - 1.8*model2.x5 + 6*model2.x6 <= 0)
model2.constr4 = Constraint(expr = 7*model2.x1 - 2*model2.x2 - 3*model2.x3 + 4*model2.x4 + 5*model2.x5 - 6*model2.x6 + 12*model2.y <= 0)

model2.constr5 = Constraint(expr = model2.y <= 5.5)
model2.constr6 = Constraint(expr = model2.y >= -0.2)


In [17]:
model2.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    y : S

In [18]:
!apt-get install -y -qq glpk-utils

In [19]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 1.10003145695364
  Upper bound: 1.10003145695364
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 8
  Number of nonzeros: 25
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.023319721221923828
# ----------------------------------------------------------
#   Solution Information
# -------------

In [20]:
print('/nExercise2 (3)')
print('\nCost = ', model2.cost())

print('\nDecision Variables')
print('x1 = ', model2.x1())
print('x2 = ', model2.x2())
print('x3 = ', model2.x3())
print('x4 = ', model2.x4())
print('x5 = ', model2.x5())
print('x6 = ', model2.x6())
print('y = ', model2.y())

print('\nConstraints')
print('Constraint 1  = ', model2.constr1())
print('Constraint 2  = ', model2.constr2())
print('Constraint 3  = ', model2.constr3())
print('Constraint 4  = ', model2.constr4())
print('Constraint 5  = ', model2.constr5())
print('Constraint 6  = ', model2.constr6())

/nExercise2 (3)

Cost =  1.1000314569536414

Decision Variables
x1 =  0.0
x2 =  19.5529801324503
x3 =  24.2384105960265
x4 =  0.0
x5 =  18.9072847682119
x6 =  0.0
y =  1.44039735099338

Constraints
Constraint 1  =  80.00000000000011
Constraint 2  =  30.00000000000002
Constraint 3  =  -9.237055564881302e-14
Constraint 4  =  -6.039613253960852e-14
Constraint 5  =  1.44039735099338
Constraint 6  =  1.44039735099338


Exercise2 (3) ->

constraints 1

3x1 − 3x2 + 5x3 − x4 + x5 − y ≤ 80 is Active 
because L.H.S is 3(0)-3(19.55)+5(24.23)-(0)+(18)-(1.44) = 80 = R.H.S

constraints 2

x3 + 2x4 + 4y ≥ 30 is Active 
because L.H.S is (24.23) + 2(0) + 4(1.44) = 30 = R.H.S

constraints 3 is inactive beacuse L.H.L $ \\ \neq $ R.H.S

constraints 4 is inactive beacuse L.H.L $ \\ \neq $ R.H.S

constraints 5 is inactive beacuse L.H.L $ \\ \neq $ R.H.S

I take two assumption in this question.
one is -x2-x4+x4 > 0 and second is -x1 + x3 + x6 -2y

after solving this problem by solver factory glpk , 

2nd assumption is satisfied but 1st assumption is not.


Exerice 3 Solution

Exercise3 (1) ->

$ \text{Let How much Kg product P produce}= y_1 \\
\ \ \ \ \ \ \text{How much Kg product Q produce} = y_2 \\ 
  \\
 \text{min} \ 1200y_1+2580y_2 \\ \text{s.t.} \ \ y_1 + y_2 >= 145 \\ 2.18y_1 + \frac{13}{12}y_2<=980 \\ y_1 >= 105 \\ y_2>=125 $

In [21]:
!pip install -q pyomo

In [22]:
from pyomo.environ import *

In [23]:
model3 = ConcreteModel()

In [24]:
model3.y1 = Var(domain=Reals)
model3.y2 = Var(domain=Reals)

In [25]:
model3.objective = Objective(expr = 1200*model3.y1+2580*model3.y2,sense=minimize)

In [26]:
model3.cons1 = Constraint(expr = model3.y1 + model3.y2 >= 145)
model3.cons2 = Constraint(expr = 2.18*model3.y1+(13/12)*model3.y2<=11760)
model3.cons3 = Constraint(expr = model3.y1>=105)
model3.cons4 = Constraint(expr = model3.y2>=125)

In [27]:
model3.pprint()

2 Var Declarations
    y1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 1200*y1 + 2580*y2

4 Constraint Declarations
    cons1 : Size=1, Index=None, Active=True
        Key  : Lower : Body    : Upper : Active
        None : 145.0 : y1 + y2 :  +Inf :   True
    cons2 : Size=1, Index=None, Active=True
        Key  : Lower : Body                            : Upper   : Active
        None :  -Inf : 2.18*y1 + 1.0833333333333333*y2 : 11760.0 :   True
    cons3 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None : 105.0 :   y1 :  +Inf :   True
    cons4 :

In [28]:
!apt-get install -y -qq glpk-utils

In [29]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 448500.0
  Upper bound: 448500.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.013638734817504883
# ----------------------------------------------------------
#   Solution Information
# ------------------------------

In [30]:
print('/nExercise3 (2)')
print('\nCost = ', model3.objective())

print('\nDecision Variables')
print('y1 = ', model3.y1())
print('y2 = ', model3.y2())

print('\nConstraints')
print('Constraint 1  = ', model3.cons1())
print('Constraint 2  = ', model3.cons2())
print('Constraint 3  = ', model3.cons3())
print('Constraint 4  = ', model3.cons4())

/nExercise3 (2)

Cost =  448500.0

Decision Variables
y1 =  105.0
y2 =  125.0

Constraints
Constraint 1  =  230.0
Constraint 2  =  364.31666666666666
Constraint 3  =  105.0
Constraint 4  =  125.0


$Activities \ \ of  \ \ all \ \ constraints$

$Constraints 1$

$y_1 + y_2 >= 145 $ is inactive because L.H.S is 105 + 125 $ \\ \neq 145 $ = R.H.S

$Constraints 2$

$ 2.18y_1 + \frac{13}{12}y_2<=980 $ is inactive because L.H.S  $= 2.18(105) +  \frac{13}{12}(124)  \neq 980 = $ R.H.S

$Constraints 3$

$ y_1 >= 105 $ is active beacuse L.H.S = $y_1$ = 105 = R.H.S

$Constraints 4$

$y_2>=125$ is Active because L.H.S = y_2 = 125 = R.H.S

$ \textbf{ Exercise3 (4) Solution-> }  \\
$


$ \text{Let How much Kg product P produce}= y_1 \\
\ \ \ \ \ \ \text{How much Kg product Q produce} = y_2 \\ 
  \\
 \text{min} \ 1200y_1+2580y_2 \\ \text{s.t.} \ \ y_1 + y_2 >= 145 \\ 2.18y_1 + \frac{13}{12}y_2<=980 \\ y_1 >= 147 \\ y_2>=222 $

Demand of y1 and y2 changed from 105 to 147 and 125 to 222.

In [31]:
!pip install -q pyomo

In [32]:
from pyomo.environ import *

In [33]:
model4 = ConcreteModel()

In [34]:
model4.y1 = Var(domain=Reals)
model4.y2 = Var(domain=Reals)

In [35]:
model4.objective = Objective(expr = 1200*model4.y1+2580*model4.y2,sense=minimize)

In [36]:
model4.cons1 = Constraint(expr = model4.y1 + model4.y2 >= 145)
model4.cons2 = Constraint(expr = 2.18*model4.y1+(13/12)*model4.y2<=11760)
model4.cons3 = Constraint(expr = model4.y1>=147)
model4.cons4 = Constraint(expr = model4.y2>=222)

In [37]:
!apt-get install -y -qq glpk-utils

In [38]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model4).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 749160.0
  Upper bound: 749160.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.02141547203063965
# ----------------------------------------------------------
#   Solution Information
# -------------------------------

In [39]:
print('/nExercise3 (6)')

print('\nCost = ', model4.objective())

print('\nDecision Variables')
print('y1 = ', model4.y1())
print('y2 = ', model4.y2())

print('\nConstraints')
print('Constraint 1  = ', model4.cons1())
print('Constraint 2  = ', model4.cons2())
print('Constraint 3  = ', model4.cons3())
print('Constraint 4  = ', model4.cons4())

/nExercise3 (6)

Cost =  749160.0

Decision Variables
y1 =  147.0
y2 =  222.0

Constraints
Constraint 1  =  369.0
Constraint 2  =  560.96
Constraint 3  =  147.0
Constraint 4  =  222.0


$Exercise3 (6) ->$

$Activities of all constraints$

$Constraints 1$

$y_1 + y_2 >= 145 $ is inactive because L.H.S is 147 + 222 $ \\ \neq 145 $ = R.H.S

$Constraints 2$

$ 2.18y_1 + \frac{13}{12}y_2<=980 $ is inactive because L.H.S  $= 2.18(147) +  \frac{13}{12}(222)  \neq 980 = $ R.H.S

$Constraints 3$

$ y_1 >= 105 $ is active beacuse L.H.S = $y_1$ = 147 = R.H.S

$Constraints 4$

$y_2>=125$ is Active because L.H.S = y_2 = 222 = R.H.S

In jen industries receive another order from different customer, 
after incresing demad minimise the total production cost.
after the soltion, cost price increase. 
for production y1 = 147  and y2 = 222 requires minimum cost is 749160

I advise jen industries about the new requirement is for requred y1 and y2 product, minimum cost is 749160